packages/fxa-event-broker/lib/selfUpdatingService/index.ts

```diff
@@ -0,0 +1,78 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { Logger } from 'mozlog';
+
+import { isError, Result } from '../result';
+
+abstract class SelfUpdatingService<T> {
+  protected logger: Logger;
+  private data: T;
+  private timer: NodeJS.Timeout | null;
+  private readonly refreshInterval: number;
+
+  constructor(logger: Logger, refreshInterval: number, defaultValue: T) {
+    this.logger = logger;
+    this.refreshInterval = refreshInterval;
+    this.data = defaultValue;
+    this.timer = null;
+  }
+
+  /**
+   * Start the service with automatic refreshing of `T` data.
+   *
+   * * This will error on start if the first fetch fails.
+   * * Calls while the service is already running will be ignored.
+   */
+  public async start() {
+    if (this.timer) {
+      return;
+    }
+
+    // Verify we can do the initial fetch
+    const result = await this.updateFunction();
+    if (isError(result)) {
+      this.logger.error('clientCapabilityService', { err: result });
+      throw result;
+    }
+    this.data = result;
+    this.timer = setInterval(async () => await this.updateService(), this.refreshInterval);
```

Comment:
I'm just splitting hairs here so no need to change anything, but if I was implementing this I'd probably call `setTimeout` inside `updateService` instead of using `setInterval` here.

It would require an extra bit of state for sane cancellation, but would prevent multiple concurrent handlers from being spawned in the event that `this.updateFunction` is badly behaved and doesn't fulfil it's promise for some reason (i.e. potentially leaky).

Reply:
I was contemplating adding a bit more around the updateService call here to at least log exceptions in the updateFunction so that this would always wrap up correctly. I'll add that in a separate PR.